stateproof/builder.go

```diff
@@ -92,59 +211,83 @@ func (spw *Worker) initBuilders() {
 	spw.mu.Lock()
 	defer spw.mu.Unlock()
 
-	var roundSigs map[basics.Round][]pendingSig
-	err := spw.db.Atomic(func(ctx context.Context, tx *sql.Tx) (err error) {
-		roundSigs, err = getPendingSigs(tx)
-		return
-	})
+	rnds, err := spw.getAllOnlineBuilderRounds()
 	if err != nil {
-		spw.log.Warnf("initBuilders: getPendingSigs: %v", err)
+		spw.log.Errorf("initBuilders: failed to load rounds: %v", err)
 		return
 	}
 
-	for rnd, sigs := range roundSigs {
+	for _, rnd := range rnds {
 		if _, ok := spw.builders[rnd]; ok {
 			spw.log.Warnf("initBuilders: round %d already present", rnd)
 			continue
 		}
-		spw.addSigsToBuilder(sigs, rnd)
+
+		buildr, err := spw.loadOrCreateBuilderWithSignatures(rnd)
+		if err != nil {
+			spw.log.Warnf("initBuilders: failed to load builder for round %d", rnd)
+			continue
+		}
+		spw.builders[rnd] = buildr
 	}
+
+	spw.ledger.RegisterVotersCommitListener(spw)
 }
 
-func (spw *Worker) addSigsToBuilder(sigs []pendingSig, rnd basics.Round) {
-	builderForRound, err := spw.makeBuilderForRound(rnd)
+func (spw *Worker) getAllOnlineBuilderRounds() ([]basics.Round, error) {
+	// Some state proof databases might only contain a signature table. For that reason, when trying to create builders for possible state proof
+	// rounds we search the signature table and not the builder table
+	latest := spw.ledger.Latest()
+	latestHdr, err := spw.ledger.BlockHdr(latest)
 	if err != nil {
-		spw.log.Warnf("addSigsToBuilder: makeBuilderForRound(%d): %v", rnd, err)
-		return
+		return nil, err
+	}
+	proto := config.Consensus[latestHdr.CurrentProtocol]
+	if proto.StateProofInterval == 0 { // StateProofs are not enabled yet
+		return nil, err
 	}
-	spw.builders[rnd] = builderForRound
 
-	for _, sig := range sigs {
-		pos, ok := builderForRound.voters.AddrToPos[sig.signer]
-		if !ok {
-			spw.log.Warnf("addSigsToBuilder: cannot find %v in round %d", sig.signer, rnd)
-			continue
-		}
+	latestStateProofRound := latest.RoundDownToMultipleOf(basics.Round(proto.StateProofInterval))
+	threshold := onlineBuildersThreshold(&proto, latestHdr.StateProofTracking[protocol.StateProofBasic].StateProofNextRound)
 
-		isPresent, err := builderForRound.Present(pos)
-		if err != nil {
-			spw.log.Warnf("addSigsToBuilder: failed to invoke builderForRound.Present on pos %d - %v", pos, err)
-			continue
-		}
-		if isPresent {
-			spw.log.Warnf("addSigsToBuilder: cannot add %v in round %d: position %d already added", sig.signer, rnd, pos)
-			continue
-		}
+	var rnds []basics.Round
+	err = spw.db.Atomic(func(_ context.Context, tx *sql.Tx) error {
+		var err error
+		rnds, err = getSignatureRounds(tx, threshold, latestStateProofRound)
```

Comment:
yes
The slice will have gaps and that is exactly the optimization we aimed for. If state proofs are being lagged for several intervals we don't want to keep all the potential builders in memory (since there might be a lot). so keep only the earliest `buildersCacheLength` builders + the last one.  
see if this comment makes sense 
https://github.com/algorand/go-algorand/blob/760c231c91136449bf2b7222a992576bcc6e394f/stateproof/builder.go#L578